pilot/pkg/model/validation.go

```diff
@@ -1529,6 +1559,42 @@ func ValidateAuthorizationPolicy(name, namespace string, msg proto.Message) erro
 	return nil
 }
 
+// isFirstClassFieldEmpty return false if there is at least one first class field (e.g. properties)
+func isFirstClassFieldEmpty(subject *rbac.Subject) bool {
+	return len(subject.User) == 0 && len(subject.Group) == 0 && len(subject.Properties) == 0 &&
+		len(subject.Namespaces) == 0 && len(subject.NotNamespaces) == 0 && len(subject.Groups) == 0 &&
+		len(subject.NotGroups) == 0 && len(subject.Ips) == 0 && len(subject.NotIps) == 0 &&
+		len(subject.Names) == 0 && len(subject.NotNames) == 0
+}
+
+// Returns true if the user defines a property that already exists in the first-class fields, false
+// if none has overlapped.
+// In the future when we introduce expression conditions in properties field, we might need to revisit
+// this function.
+func hasExistingFirstClassFieldInBinding(propertiesKey string, subject *rbac.Subject) bool {
+	switch propertiesKey {
+	case "source.principal":
+		return len(subject.Names) > 0
+	case "request.auth.claims[groups]":
+		return len(subject.Groups) > 0
+	case "source.namespace":
+		return len(subject.Namespaces) > 0
+	case "source.ip":
+		return len(subject.Ips) > 0
+	}
+	return false
+}
+
+// isStarInNames returns true if the user defines "*" as of the values for names or not_names in ServiceRoleBinding.
```

Comment:
No need to mention the names or not_names or ServiceRoleBinding, it's not the logic of this function but more related to the caller.

Reply:
Good point. Will edit.